distributed/worker.py

```diff
@@ -3547,39 +3571,60 @@ def dumps_task(task):
     >>> dumps_task(1)  # doctest: +SKIP
     {'task': b'\x80\x04\x95\x03\x00\x00\x00\x00\x00\x00\x00K\x01.'}
     """
-    if istask(task):
+    if not check_task or istask(task):
         if task[0] is apply and not any(map(_maybe_complex, task[2:])):
-            d = {"function": dumps_function(task[1]), "args": warn_dumps(task[2])}
+            # Use `warn_serialize` if any elements of `task` are `Serialized`.
+            # Otherwise, we can just use warn_dumps (pickle)
+            _dumps = warn_serialize if _serialized_task_arg(task[2:]) else warn_dumps
+            d = {"function": dumps_function(task[1]), "args": _dumps(task[2])}
```

Comment:
@gjoseph92 @ian-r-rose @jrbourbeau - The changes in this PR seem to address the requirements in #4673 (in fact, nested tasks also work), but I am not happy with the changes in `dumps_task`.

Note that `_serialized_task_arg` is checking the type of each (non-function) element in `task`, and using `warn_serialize` rather than `warn_dumps` if there are any `Serialized` objects.  This extra pass over the data partially defeats the purpose of the check (we might as well always use `warn_serialize` if we are going to take the time to iterate over elements anyway).  Perhaps we should introduce a standalone `serialize_task` function that will do the same thing as `dumps_task`, but will always call `warn_serialize` on the args and kwargs?  In that case, the decision whether to use `dumps_task` vs `serialize_task` would need to be made upstream, but the logic in `dumps_task` could remain the same.

I guess we could also cache the _serialized_task_arg result based on the function, but that seems like it could be fragile.

Reply:
Oaky - I think I'd like to leave `dumps_task` alone and introduce the new logic for serializing a task on the scheduler in a new `serialize_task` function (I introduced this in 093d240)